### ISSUE TITLE
Rule 'indent' complains on multi-line function definition

### ISSUE THREAD

**Issue by torarvid** [eslint/eslint#2613]
Hi! Sometimes, I will write a unit test like

    it('Should do blah blah with bluh bluh and this line is '
        + 'very long so I break it up to two lines', function() {
    
        doStuff(); // <-- will complain about indentation here
    });

.. and eslint complains on the `doStuff` line (it wants another level of indentation). However, if the test description is short enough to fit on a single line like this:

    it('Should do blah', function() {
    
        doStuff(); // <-- no complaints about indentation
    });

Then everything is ok. To me, the fact that the test description is long (and breaks to the next line) shouldn't change how much the function contents should be indented.

**Comment by gyandeeps** [comment]
Its a know issue. 
Related https://github.com/eslint/eslint/issues/1801

**Comment by ilyavolodin** [comment]
Closing in favor of #1801

**Comment by lo1tuma** [comment]
I'm not sure if this is a duplicate of #1801.
#1801 is about missing warnings on multiline expressions, but this issue is about a false extra warning **after** a multiline expression.

**Comment by ilyavolodin** [comment]
Ahh, sorry, my bad, reopening.

**Comment by ilyavolodin** [comment]
Example above returns two errors:
```
4:8 - Expected indentation of 8 characters.
5:4 - Expected indentation of 4 characters.
```
I'm actually not 100% sure it's incorrect behavior. Since the opening bracket of the function is indented by 4 spaces, the body of the function should be indented by 8 (which eslint rightfully complains about). And the second error is for closing of function's body, which should be at the same indentation as opening brace (so 4 spaces).
Somebody who is more familiar with indent rule, correct me if I'm wrong, but I'm not seeing an issue here.

**Comment by torarvid** [comment]
Hmm, first I thought you were not thinking clearly, @ilyavolodin - but now I think maybe **I** wasn't :wink: 

Because you're saying that the right way to indent might be:

    it('long string',
        + ' blah blablah', function() {
        
            doStuff();
        }
    );

.. right? It feels like it's about the same as:

    it('long string',
        + ' blah blablah'
        , function() {        
            doStuff();
        }
    );

(I haven't tested but if these two have no errors, I'm inclined to agree that I might have mis-reported this issue :smile:)

**Comment by ilyavolodin** [comment]
@torarvid Both of the examples you provided do not error with indent on. So I think this is a non-issue.

**Comment by ilyavolodin** [comment]
Closing, as this seems to behave correctly.

**Comment by okhomenko** [comment]
Let's say

```javascript
/*eslint "indent": [2, 4] */

function MyController($scope,
                           $rootScope) {
    var vm = this; // 5:5  error  Expected indentation of 26 characters but found 4
}
```
```sh
$ eslint --version
v1.1.0
```

Not sure that it behaves correctly.

**Comment by ilyavolodin** [comment]
@okhomenko Can you open a new issue for that error?

**Comment by gyandeeps** [comment]
That is multi-line, which is an known issue. we already have an issue open for that.

**Comment by leizhao4** [comment]
@torarvid I think you reported this correctly. I don't agree with your second comment, which implied that parentheses should increase indentation level. It shouldn't. If they do, we will have very ugly code looks like this:
```
console.log(a * (b + (c -
            d // Indentation level 3?
        )
    )
);
```
or
```
console.log(a * (b + (c -
            d))); // Indentation level 3?
```

It doesn't look right to me. How about this?
```
console.log(a * (b + (c -
    d))); // Indentation level 1 :-)
```

Much better, right? That means, the indentation level increase does not come from the parentheses. It comes from the line break.

So the next question, should the indentation level increase (caused by line breaks) be temporary (only effective on the next one line), or permanent (effective on all following lines, until the next indentation level decrease)? I would say temporary. Because a permanent increase will give us ugly code like this:
```
console.log('This is a very very very very' +
    ' very very very very' +
        ' very very very very' +
            ' very very very very' +
                ' very long line that breaks.');
```

That doesn't look right to me. Also, unlike other symbols that increase indentation level, a line break doesn't have a closing symbol. If it is a permanent increase, when do you decrease?

So the only reasonable answer is, it should be temporary. It should only increase indentation level of the next one line, and should not affect any lines after that. Like this:
```
console.log('This is a very very very very' +
    ' very very very very' + // Line 2 has an indentation level 1 caused by line 1
    ' very very very very' + // Line 3 has an indentation level 1 caused by line 2
    ' very very very very' + // Line 4 has an indentation level 1 caused by line 3
    ' very long line that breaks.'); // and so on
```
or
```
it('Should do blah blah with bluh bluh and this line is '
    + 'very long', function() { // Indentation level 1 because of line break
    doStuff(); // Indentation level 1 because of curly braces
    // Line break after line 1 should only affect line 2, not 3 and 4.
});
```

@ilyavolodin What do you think? :-)

**Comment by ilyavolodin** [comment]
@leizhao4 I personally agree with all the cases you posted, except the last one. In that one `doStuff()` should be indented two levels, same with the comment following it. While we would love to support every possible case, indent is by far the most complicated rule that we have (out of 200+ rules), and it's in a bad state currently.

**Comment by leizhao4** [comment]
Hmm, I start to get confused because when I try the same code in ESLint Demo, I get different results, which is my expected behavior.

![screen shot 2016-05-23 at 8 56 54 pm](https://cloud.githubusercontent.com/assets/763847/15489887/29160188-2129-11e6-8f02-0ede1c64a7b4.png)

**Comment by ilyavolodin** [comment]
All of those issues that you listed are related to multi-line statements, which is a known issue with indent.
